src/JsonSchema/Constraints/SchemaConstraint.php

```diff
@@ -20,23 +23,70 @@
  */
 class SchemaConstraint extends Constraint
 {
+    const DEFAULT_SCHEMA_SPEC = 'http://json-schema.org/draft-04/schema#';
+
     /**
      * {@inheritdoc}
      */
     public function check(&$element, $schema = null, JsonPointer $path = null, $i = null)
     {
         if ($schema !== null) {
             // passed schema
-            $this->checkUndefined($element, $schema, $path, $i);
+            $validationSchema = $schema;
         } elseif ($this->getTypeCheck()->propertyExists($element, $this->inlineSchemaProperty)) {
             $inlineSchema = $this->getTypeCheck()->propertyGet($element, $this->inlineSchemaProperty);
             if (is_array($inlineSchema)) {
                 $inlineSchema = json_decode(json_encode($inlineSchema));
             }
             // inline schema
-            $this->checkUndefined($element, $inlineSchema, $path, $i);
+            $validationSchema = $inlineSchema;
         } else {
             throw new InvalidArgumentException('no schema found to verify against');
         }
+
+        // validate schema against whatever is defined in $validationSchema->$schema. If no
+        // schema is defined, assume self::DEFAULT_SCHEMA_SPEC (currently draft-04).
+        if ($this->factory->getConfig(self::CHECK_MODE_VALIDATE_SCHEMA)) {
+            if (!$this->getTypeCheck()->isObject($validationSchema)) {
+                throw new RuntimeException('Cannot validate non-object schema');
+            }
+            if ($this->getTypeCheck()->propertyExists($validationSchema, '$schema')) {
+                $schemaSpec = $this->getTypeCheck()->propertyGet($validationSchema, '$schema');
+            } else {
+                $schemaSpec = self::DEFAULT_SCHEMA_SPEC;
+            }
+
+            // preload standard schema specs
+            $preload = array(
+                'http://json-schema.org/draft-03/schema' => 'json-schema-draft-03.json',
+                'http://json-schema.org/draft-03/schema#' => 'json-schema-draft-03.json',
+                'http://json-schema.org/draft-04/schema' => 'json-schema-draft-04.json',
+                'http://json-schema.org/draft-04/schema#' => 'json-schema-draft-04.json'
+            );
+            $schemaStorage = $this->factory->getSchemaStorage();
+            foreach ($preload as $schemaID => $schemaFile) {
+                $schemaStorage->addSchema(
+                    $schemaID,
+                    json_decode(file_get_contents(__DIR__ . "/../../../schema-validation/$schemaFile"))
```

Comment:
No. The problem is the inability to use scalar expressions < 5.6, which means we have to use a different mechanism. We don't know what the working directory is, which means we need to use a path relative to `__DIR__` (your example is missing this). Using `__DIR__` implies string concatenation, which implies an expression.

What we're trying to achieve is some kind of global mechanism that provides the root filesystem path of the json-schema package, for further use by whatever else might need to know that, rather than manually constructing the path wherever it's used.